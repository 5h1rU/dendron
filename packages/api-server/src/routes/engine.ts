import {
  DEngineV2,
  EngineQueryRequest,
  EngineWriteRequest,
} from "@dendronhq/common-all";
import { DendronEngine } from "@dendronhq/engine-server";
import { Request, Response, Router } from "express";
import { MemoryStore } from "../store/memoryStore";

const router = Router();

router.get("/health", async (_req: Request, res: Response) => {
  return res.json({ ok: 1 });
});

router.post("/query", async (req: Request, res: Response) => {
  const { ws, queryString, mode } = req.body as EngineQueryRequest;
  const engine = await MemoryStore.instance().get<DendronEngine>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  const { error, data } = await engine.query(queryString, mode);
  const cleanData = data.map((ent) => ent.toRawProps());
  res.json({ error, data: cleanData });
});

router.post("/write", async (req: Request, res: Response) => {
  const { ws, node, opts } = req.body as EngineWriteRequest;
  const engine = await MemoryStore.instance().get<DEngineV2>(`ws:${ws}`);
  if (!engine) {
    throw "No Engine";
  }
  try {
    await engine.writeNote(node, opts);
    res.json({ error: null, data: null });
  } catch (err) {
    res.json({ error: JSON.stringify(err), data: null });
  }
});

export { router as engineRouter };
