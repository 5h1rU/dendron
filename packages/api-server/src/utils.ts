import { DEngineV2 } from "@dendronhq/common-all/src";
import { DendronEngineV2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { MemoryStore } from "./store/memoryStore";

export function getWSKey(uri: string) {
  return _.trimEnd(uri, "/");
}

export async function putWS({
  ws,
  engine,
}: {
  ws: string;
  engine: DendronEngineV2;
}) {
  MemoryStore.instance().put(`ws:${getWSKey(ws)}`, engine);
}

export async function getWS({ ws }: { ws: string }) {
  const engine = await MemoryStore.instance().get<DEngineV2>(
    `ws:${getWSKey(ws)}`
  );
  if (!engine) {
    throw `No Engine: ${ws}`;
  }
  return engine;
}
