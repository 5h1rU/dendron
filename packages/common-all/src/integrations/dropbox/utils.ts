import { Note } from "../../node";
import { files } from "dropbox";
import path from "path";

export function denddronId2DxId(did: string): string {
  return `id:${did}`;
}
export function dxId2DendronId(dxId: string): string {
  return dxId.slice(3);
}

export function fileNameToTitle(name: string): string {
  const ext = path.extname(name);
  return path.basename(name, ext);
}

export function fileNameToTreePath(name: string): string[] {
  const title = fileNameToTitle(name);
  return title.split(".");
}

export function fileToNote(
  meta:
    | files.FileMetadata
    | files.FileMetadataReference
    | files.FolderMetadataReference,
  body?: string
) {
  const title = fileNameToTitle(meta.name);
  const id = dxId2DendronId(meta.id);
  const note = new Note({
    id,
    title,
    desc: "TODO",
    type: "note",
    schemaId: "-1",
    body
  });
  return note;
}
