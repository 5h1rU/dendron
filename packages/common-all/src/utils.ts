import minimatch from "minimatch";

export class DUtils {
  static minimatch = minimatch;
}
