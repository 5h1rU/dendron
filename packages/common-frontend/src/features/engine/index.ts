export * as engineReducer from "./slice";
export {
  setError,
  setNotes,
  EngineSliceUtils,
  initNotes,
  InitNoteOpts,
  EngineState,
} from "./slice";
export * as engineStore from "./store";
export {
  RootState as engineState,
  AppDispatch as engineDispatch,
} from "./store";
