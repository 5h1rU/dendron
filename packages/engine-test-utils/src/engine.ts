import { DEngineClientV2, DVault, WorkspaceOpts } from "@dendronhq/common-all";
import { launch } from "@dendronhq/api-server";
import { tmpDir } from "@dendronhq/common-server";
import {
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
} from "@dendronhq/common-test-utils";
import {
  createEngine as engineServerCreateEngine,
  DendronEngineClient,
  WorkspaceService,
} from "@dendronhq/engine-server";
import _ from "lodash";

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<DEngineClientV2>;

export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return engineServerCreateEngine(opts);
}

export { DEngineClientV2, DVault, WorkspaceOpts };

export async function createEngineFromServer(opts: WorkspaceOpts) {
  const port = await launch({});
  const engine: DEngineClientV2 = DendronEngineClient.create({
    port,
    vaultsv4: opts.vaults,
    ws: opts.wsRoot,
    vaults: [],
  });
  await engine.init();
  return engine;
}

async function setupWS(opts: { vaults: DVault[] }) {
  const wsRoot = tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  const vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault });
      return vault;
    })
  );
  return { wsRoot, vaults };
}

// @ts-ignore
export async function runEngineTestV5<TExtra = any>(
  func: RunEngineTestFunctionV4,
  opts: {
    preSetupHook?: SetupHookFunction;
    //postSetupHook?: PostSetupHookFunction;
    createEngine?: AsyncCreateEngineFunction;
    extra?: TExtra;
    expect: any;
    vaults?: DVault[];
    setupOnly?: boolean;
  }
): Promise<any> {
  const { preSetupHook, extra, vaults, createEngine } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    vaults: [{ fsPath: "vault1" }, { fsPath: "vault2" }],
  });
  const { wsRoot } = await setupWS({ vaults });
  await preSetupHook({ wsRoot, vaults });
  const engine = await createEngine({ wsRoot, vaults });
  const initResp = await engine.init();
  const testOpts = { wsRoot, vaults, engine, initResp, extra };
  if (opts.setupOnly) {
    return testOpts;
  }
  const results = (await func(testOpts)) || [];
  await runJestHarnessV2(results, expect);
  return { opts: testOpts, resp: undefined };
}
