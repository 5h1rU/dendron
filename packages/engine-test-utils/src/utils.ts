import { AssertUtils } from "@dendronhq/common-test-utils";
import { Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }
  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
