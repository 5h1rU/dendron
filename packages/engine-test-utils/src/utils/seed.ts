import { DEngineClient, SeedCommands, SeedConfig } from "@dendronhq/common-all";
import { tmpDir, writeYAML } from "@dendronhq/common-server";
import { SeedCLICommand } from "@dendronhq/dendron-cli";
import { SeedInitMode, SeedUtils } from "@dendronhq/engine-server";
import path from "path";
import { GitTestUtils } from "./git";

export class TestSeedUtils {
  static async createSeedRegistry(opts: {
    engine: DEngineClient;
    wsRoot: string;
  }) {
    const config = await this.createSeed(opts);
    const id = SeedUtils.getSeedId(config);
    const root = tmpDir().name;
    const registryFile = path.join(root, "reg.yml");
    writeYAML(registryFile, { [id]: config });
    return { registryFile };
  }
  static async createSeed(opts: { engine: DEngineClient; wsRoot: string }) {
    const cli = new SeedCLICommand();
    const cmd = SeedCommands.INIT;
    const id = "dendron.foo";
    const seed: SeedConfig = {
      description: "",
      license: "",
      name: "foo",
      publisher: "dendron",
      repository: {
        type: "git",
        url: `file://${opts.wsRoot}`,
      },
      root: "vault",
    };
    await cli.execute({
      cmd,
      id,
      server: {} as any,
      config: seed,
      mode: SeedInitMode.CREATE_WORKSPACE,
      ...opts,
    });
    await GitTestUtils.addRepoToWorkspace(opts.wsRoot);
    return seed;
  }
}
