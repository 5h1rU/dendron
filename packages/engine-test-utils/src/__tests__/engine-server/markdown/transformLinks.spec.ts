import { DEngineClient, DNoteLoc, DVault } from "@dendronhq/common-all";
import {
  DendronASTDest,
  MDUtilsV4,
  TransformLinkOpts,
  transformLinks,
} from "@dendronhq/engine-server";
import { testWithEngine } from "../../../engine";
import { checkVFile } from "./utils";

function cproc(
  opts: { engine: DEngineClient; vault: DVault } & TransformLinkOpts
) {
  const { engine, vault, from, to } = opts;
  const proc = MDUtilsV4.procFull({
    dest: DendronASTDest.MD_DENDRON,
    engine,
    vault,
    fname: "foo",
  }).use(transformLinks, { from, to });
  return proc;
}

describe("replaceLinks", () => {
  describe("single", () => {
    testWithEngine("basic", async ({ engine, vaults }) => {
      const vault = vaults[0];
      const content = ["[[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
      };
      const to: DNoteLoc = {
        fname: "bar",
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "[[bar]]");
    });

    testWithEngine("ref", async ({ vaults, engine }) => {
      const vault = vaults[0];
      const content = ["![[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "![[bar]]");
    });

    testWithEngine("multiple", async ({ engine, vaults }) => {
      const vault = vaults[0];
      const content = ["[[bond]] [[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "[[bar]] [[bar]]");
    });

    testWithEngine("inline code", async ({ vaults, engine }) => {
      const vault = vaults[0];
      const content = ["`[[bond]]`"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "`[[bond]]`");
    });

    testWithEngine("fenced code", async ({ vaults, engine }) => {
      const vault = vaults[0];
      const content = ["```", "[[bond]]", "```"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "```\n[[bond]]\n```");
    });

    testWithEngine("with alias", async ({ vaults, engine }) => {
      const content = ["[[hero|bond]]"].join("\n");
      const vault = vaults[0];
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "[[hero|bar]]");
    });

    testWithEngine("with offset ", async ({ vaults, engine }) => {
      const vault = vaults[0];
      const content = ["   [[bond]]"].join("\n");
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "[[bar]]");
    });

    testWithEngine("with vault prefix", async ({ vaults, engine }) => {
      const vault = vaults[0];
      const content = "[[dendron://vault1/bond]]";
      const from: DNoteLoc = {
        fname: "bond",
        vault,
      };
      const to: DNoteLoc = {
        fname: "bar",
        vault,
      };
      const resp = await cproc({ engine, vault, from, to }).process(content);
      await checkVFile(resp, "[[dendron://vault1/bar]]");
    });
  });
});
