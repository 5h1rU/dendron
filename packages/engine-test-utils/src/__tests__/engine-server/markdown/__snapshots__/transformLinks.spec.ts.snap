// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceLinks single basic 1`] = `
VFile {
  "contents": "[[bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single fenced code 1`] = `
VFile {
  "contents": "\`\`\`
[[bond]]
\`\`\`

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single inline code 1`] = `
VFile {
  "contents": "\`[[bond]]\`

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single multiple 1`] = `
VFile {
  "contents": "[[bar]] [[bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single ref 1`] = `
VFile {
  "contents": "![[bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single with alias 1`] = `
VFile {
  "contents": "[[hero|bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single with offset  1`] = `
VFile {
  "contents": "   [[bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`replaceLinks single with vault prefix 1`] = `
VFile {
  "contents": "[[dendron://vault1/bar]]

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
